static-analysis/teammates-mackerTest.xml

```diff
@@ -0,0 +1,159 @@
+        <access-rule>
+            <deny>
+                <to class="teammates.client.remoteapi.**" />
+                <allow>
+                    <from class="teammates.client.scripts.**" />
+                </allow>          
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="packageDiagram: client::util should only be used inside client">
+        <access-rule>
+            <deny>
+                <to class="teammates.client.scripts.util.**" />
+                <allow>
+                    <from class="teammates.client.**" />
+                </allow>          
+            </deny>
+        </access-rule>
+    </ruleset>
+
+    <ruleset name="TestDriverComponent: Test cases should not be dependent on each other">
+        <var name="testCasesPackage" value="teammates.test.cases" />
+        <access-rule>
+            <deny>
+                <from class="${testCasesPackage}.**" />
+                <to class="${testCasesPackage}.**" />
+                <allow>
+                    <!-- allow dependency from inner class -->
+                    <from class="${testCasesPackage}.**$**" />
+                </allow>
+                <allow>
+                    <to>
+                        <!-- allow dependency to inner class -->
+                        <include class="${testCasesPackage}.**$**" />
+                        <!-- allow dependency to base class -->
+                        <include class="${testCasesPackage}.**Base**" />
+                    </to>
+                </allow>
+                <allow>
+                    <!-- allow dependency from Feedback*QuestionUiTest to FeedbackQuestionUiTest -->
+                    <from class="${testCasesPackage}.browsertests.Feedback*QuestionUiTest" />
+                    <to class="${testCasesPackage}.browsertests.FeedbackQuestionUiTest" />
+                </allow>
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="TestDriverComponent: Only UI tests can access page object classes">
+        <access-rule>
+            <deny>
+                <from class="teammates.test.**" />
+                <to class="teammates.test.pageobjects.**" />
+                <allow>
+                    <from>
+                        <include class="teammates.test.pageobjects.**" />
+                        <include class="teammates.test.cases.browsertests.**" />
+                    </from>
+                </allow>
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="TestDriverComponent: Test driver can only be accessed from test package">
+        <access-rule>
+            <deny>
+                <from class="teammates.**" />
+                <to class="teammates.test.driver.**" />
+                <allow>
+                    <from class="teammates.test.**" />
+                </allow>
+                <allow>
+                    <from class="teammates.client.**" />
```

Comment:
(GitHub bug, I cannot reply the previous review, I reply here)
Put them into one file will make most of the architectural rules for production code to have `<allow>` for test case reference. I want to separate them into two files so that `mackerMain` will analysis against `main` code while `mackerTest` will analysis again the whole project.